examples/lox/src/language/type-system/lox-type-checking.ts

```diff
@@ -0,0 +1,287 @@
+            isPrintStatement,
+            (node: unknown) => isReturnStatement(node) && node.value === undefined
+        ] });
+    const typeNil = primitiveKind.createPrimitiveType({ primitiveName: 'nil',
+        inferenceRules: isNilLiteral }); // TODO for what is this used?
+    const typeAny = anyKind.createTopType({});
+
+    // utility function to map language types to Typir types
+    // TODO get rid of these type dispatch!
+    function mapType(typeRef: TypeReference): Type {
+        if (!typeRef) {
+            throw new Error('a type reference must be given');
+        }
+        if (typeRef.primitive) {
+            switch (typeRef.primitive) {
+                case 'number': return typeNumber;
+                case 'string': return typeString;
+                case 'boolean': return typeBool;
+                case 'void': return typeVoid;
+                default: assertUnreachable(typeRef.primitive);
+            }
+        } else if (typeRef.reference && typeRef.reference.ref) {
+            // search for an existing class
+            const classType = classKind.getClassType(typeRef.reference.ref.name);
+            if (classType) {
+                return classType;
+            } else {
+                throw new Error();
+            }
+        } else {
+            // search for an existing function
+            // TODO lambda vs function
+            const functionType = functionKind.getFunctionType({
+                functionName: 'TODO',
+                inputParameters: [], // TODO
+                outputParameter: undefined, // TODO
+            });
+            if (functionType) {
+                return functionType;
+            } else {
+                throw new Error();
+            }
+        }
+    }
+
+    // extract inference rules, which is possible here thanks to the unified structure of the Langium grammar (but this is not possible in general!)
+    const binaryInferenceRule: InferOperatorWithMultipleOperands<BinaryExpression> = {
+        filter: isBinaryExpression,
+        matching: (node: BinaryExpression, name: string) => node.operator === name,
+        operands: (node: BinaryExpression, _name: string) => [node.left, node.right],
+    };
+    const unaryInferenceRule: InferOperatorWithSingleOperand<UnaryExpression> = {
+        filter: isUnaryExpression,
+        matching: (node: UnaryExpression, name: string) => node.operator === name,
+        operand: (node: UnaryExpression, _name: string) => node.value,
+    };
+
+    // binary operators: numbers => number
+    for (const operator of ['-', '*', '/']) {
+        operators.createBinaryOperator({ name: operator, signature: { left: typeNumber, right: typeNumber, return: typeNumber }, inferenceRule: binaryInferenceRule });
+    }
+    operators.createBinaryOperator({ name: '+', signature: [
+        { left: typeNumber, right: typeNumber, return: typeNumber },
+        { left: typeString, right: typeString, return: typeString },
```

Comment:
As described in the comments below, the alternative types should be handled by type implicit conversions.

Reply:
I don't know, which implicit conversions are defined for LOX

Reply:
We will create an example how to work with implicit conversions.